core/pubnub_retry_configuration.c

```diff
@@ -0,0 +1,470 @@
+/**
+ * @brief Handle user-provided excluded endpoints for retry configuration.
+ *
+ * @param configuration     Pointer to the configuration for which list of
+ *                          exclusions should be set.
+ * @param excluded_endpoint First entry in the list of excluded endpoints.
+ * @param endpoints         Variadic list of other excluded endpoints.
+ */
+static void pubnub_retry_configuration_set_excluded_(
+    pubnub_retry_configuration_t* configuration,
+    pubnub_endpoint_t             excluded_endpoint,
+    va_list                       endpoints);
+
+/**
+ * @brief Check whether `transaction` can be retried after failure.
+ *
+ * @param configuration Pointer to the configuration with failed request
+ *                      handling details.
+ * @param transaction   Failed transaction type.
+ * @param current_retry How many consequential retry attempts has been done.
+ * @param http_status   Failed transaction status code.
+ * @return `true` if transaction can be handled once more after retry.
+ */
+static bool pubnub_retry_configuration_retryable_(
+    const pubnub_retry_configuration_t* configuration,
+    enum pubnub_trans                   transaction,
+    int                                 current_retry,
+    int                                 http_status);
+
+/**
+ * @brief Identify API endpoint group from transaction type.
+ *
+ * @param trans Transaction which should be translated.
+ * @return One of known PubNub API endpoint group type.
+ */
+static pubnub_endpoint_t pubnub_endpoint_from_transaction_(
+    enum pubnub_trans trans);
+
+
+// ----------------------------------------------
+//                   Functions
+// ----------------------------------------------
+
+void pubnub_set_retry_configuration(
+    pubnub_t*                     pb,
+    pubnub_retry_configuration_t* configuration)
+{
+    PUBNUB_ASSERT(pb_valid_ctx_ptr(pb));
+    pubnub_mutex_lock(pb->monitor);
+    if (NULL != pb->core.retry_configuration)
+        pubnub_retry_configuration_free(&pb->core.retry_configuration);
+    pb->core.retry_configuration = configuration;
+    pubnub_mutex_unlock(pb->monitor);
+}
+
+pubnub_retry_configuration_t* pubnub_retry_configuration_linear_alloc(void)
+{
+    return pubnub_retry_configuration_linear_alloc_with_excluded(-1);
+}
+
+pubnub_retry_configuration_t*
+pubnub_retry_configuration_linear_alloc_with_excluded(
+    const int excluded,
+    ...)
+{
+    va_list args;
+    va_start(args, excluded);
+    pubnub_retry_configuration_t* configuration =
+        pubnub_retry_configuration_alloc_with_options_(
+            PUBNUB_LINEAR_RETRY_POLICY,
+            DEFAULT_MINIMUM_DELAY,
+            0,
+            DEFAULT_LINEAR_RETRY_ATTEMPTS,
+            excluded,
+            args);
+    va_end(args);
+
+    return configuration;
+}
+
+pubnub_retry_configuration_t*
+pubnub_retry_configuration_linear_alloc_with_options(
+    const int delay,
+    const int maximum_retry,
+    const int excluded,
+    ...)
```

Comment:
Do you know what compilers are supporting this syntax? 

I don't know and I'm thinking if any of our current customers might be influenced by this change and make their build fails.

Reply:
This is the same notation as `printf` has, which I believe supported in most compilers which support the standard C library.